### PR TITLE
feat: add ui to configuration data exchanges [DHIS2-16434]

### PR DESCRIPTION
**Pull request by tomzemp**
See ticket: https://dhis2.atlassian.net/jira/software/c/projects/DHIS2/issues/DHIS2-16434
And figma mockup: https://www.figma.com/proto/7pn74aKmqiaiK2Y52XjwGN/Aggregate-Data-Exchange-Configurator

Note if you are testing, you need to use `system` user (or another user with `ALL` authority or ADD/DELETE authorities for data exchanges)

### UI Differences from figma
- id scheme information is hidden in advanced section (discussed with Joe, but we did not update the prototypes)
- we have an extra warning + click to edit if you are editing anything within `target` for an external type exchange (this is because you currently need to update authentication details when making these edits)
- if you open a exchange with inputIdScheme:CODE, you will get a warning that the app will convert to using inputIdScheme:UID on save (in general, we do not support defining with code when defining these exchanges in the app)

### Technical implementation
- Since the mockup is structured essentially as two forms (one for saving exchanges, one for saving requests), I have emulated this and nested the request form inside the exchange form (though I suppose it would be possible to structure otherwise). I considered if requests object should be a form value on exchange form but opted instead to just keep track of requests state with useReducer. The main (exchange form) which has most of the technical points of note is https://github.com/dhis2/data-exchange-app/blob/DHIS2-16434/configurations-ui/src/components/edit/exchange-update/edit-exchange.js
- Requests are stored internally and will only be committed to the back end when saving an exchange. We discussed the possibility of persisting to the backend on request edit, but this would not be possible for new exchanges (i.e. the exchange needs to already exist such that you can save a request against it).
- I've updated the folder structure to create pages and separate components between view/edit mode. I'm not really attached to where anything is, so I could do something else.
- OrgUnit, Period, and Data Item selectors are from analytics. In order to disable calculations in the Data Item selector, I am using an alpha release of analytics. 
- Filters implementation is taken from the maps app (it's not exported, so I've just copy/pasted the code and done some minor adjustments, including changing waste bin icon to remove button (to be consistent with Joe's design)
- the visualization open dialog is currently copied into the app from analytics (OpenFileDialog) because analytics library is exporting the FileMenu component which display `File` and lets you click to open the file dialog, which is not consistent with the use case here. I think it would be better to work with analytics team to export this from analytics, but there's a couple of other modifications we also will need within this app, so it seemed to just import the code directly for now.

### Known issues
- You cannot currently edit an external exchange without providing the authentication details again. This is a backend issue relating to problems with the patch update: https://dhis2.atlassian.net/browse/DHIS2-16440.
- There are not tests at the moment (I'm thinking it would make sense to come back to write tests after the soft freeze as this a continuous release app).
- David and Joe and I discussed adding the option to add a request from an existing visualization (from add/edit exchange page). I have not added this yet as there turned out to be some complexities in interpreting some of the visualization parameters into the aggregate data exchange format (e.g. when certain dimensions are filters in a visualization but need to be explicit dimensions in an aggregate data exchange). There are not big obstacles here, but I thought it would be better to submit the PR and create a ticket for the additional work on creating requests from visualizations
- Data dimension selector from analytics currently allows for calculations select by default, but we do not want this enabled. Analytics team is going to soon release a new version of the component that allows enabling/disabling certain items, so I am just using the default selector until new one is implemented.

### CODE REVIEW

File: src/components/edit/shared/form-subsection.js
Comment (Mohammer5) [inline src/components/edit/shared/form-subsection.js:14]: Needs `cursor: pointer;`

**Review by Mohammer5 — CHANGES_REQUESTED** [review]
I'll review this in multiple attempts and share what I found so I don't have to keep the tab open for a long time and potentially lose everything

I created a configuration with a single request ("Foo bar baz" on [play/dev](https://debug.dhis2.org/dev)). When I edit that configuration and delete the request, the request disappears and I can successfully save the configuration. When I then open the configuration again, the request is still there:

Peek 2024-02-13 14-10

File: src/components/edit/exchange-update/edit-exchange.js
Comment (Mohammer5) [inline src/components/edit/exchange-update/edit-exchange.js:34]: This name is misleading as it's also used by the add-view, maybe `ExchangeForm` or something along those lines is better.

**Review by Mohammer5 — COMMENTED** [review]
The components in `src/pages` import directly from the structure inside `src/components`. This is something I'd argue should be done through a consisten use of `index.js` so pages have to import from `src/components` only

File: src/components/edit/exchange-update/edit-exchange.js
Comment (tomzemp) [reply src/components/edit/exchange-update/edit-exchange.js]: Updated this name and other similar names, now using ExchangeForm, ExchangeFormContents, RequestForm, RequestFormContents

**Comment by tomzemp** [comment]
Thanks for the review @Mohammer5 🙏!

I've updated to address your points.

**Review by Mohammer5 — APPROVED** [review]
I didn't find anything obvious anymore, looks good to me so far.
Maybe it'd make sense to get another review on this one as it's so much code that it's quite easy to overlook things. I'll leave that up to you

File: src/components/edit/exchange-update/exchange-form-contents.js
Comment (amcgee) [inline src/components/edit/exchange-update/exchange-form-contents.js:119]:
```suggestion
                                        'Send data to another DHIS2 instance or data warehouse'
```

Took a screenshot of the config screen for a presentation and noticed this typo!

File: src/components/edit/exchange-update/requests-overview.js
Comment (amcgee) [reply src/components/edit/exchange-update/requests-overview.js:61]: @cooper-joe I'm curious... do we usually try to avoid referring to interface elements like this?

Comment (tomzemp) [reply src/components/edit/exchange-update/requests-overview.js]: I don't know if we generally have a policy, but we sparingly make use of language like this in some other apps (e.g. in [data entry app](https://github.com/dhis2/aggregate-data-entry-app/blob/master/src/data-workspace/entry-screen.js#L53-L57), or in the [users app](https://github.com/dhis2/user-app/blob/master/src/pages/UserList/UserTable.js#L60) [not explicitly mentioning button, but we say "below" and the button that they're supposed to use is right below))

Comment (cooper-joe) [reply src/components/edit/exchange-update/requests-overview.js]: Generally, yes, we try to avoid some types of this as it's a form of UI-layout-dependency that can become out of sync with the actual UI. However, it can be reworded slightly to remove the directionality and therefore to avoid the dependency, for example:
`Click the 'Add request' button to add a request to this exchange`.

**Review by kabaros — APPROVED** [review]
approved - could you please add a ticket to look at the tests after hard freeze (also generally the value of using cypress fixtures)

**Pull request merged**
